### PR TITLE
Simple non-automaton impl for StrField prefix queries

### PR DESCRIPTION
**Pull request by magibney**
See: [SAI-4258](https://fullstory.atlassian.net/browse/SAI-4258)

Simple prefix query implementation that avoids Automaton -- faster in all cases afaict, and _way_ faster for prefixes that match a large proportion of terms in the index (e.g., ~3x faster for prefix "https://app.fullstory.com/" on staging). I think this should ideally go upstream to Lucene in some form, but deployment would be trivial locally in the meantime -- very little potential for merge conflicts.

Should also avoid the stack-overflow errors on longer prefix strings.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/query/DynamicComplementPrefixQuery.java [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+public class DynamicComplementPrefixQuery extends MultiTermQuery {
+
+  private final Term termPrefix;
+  private final BytesRef prefix;
+  private final BytesRef limit;
+  private final MultiTermQuery inverted;
+
+  /** Constructs a query for terms starting with <code>prefix</code>. */
+  public DynamicComplementPrefixQuery(Term prefix, boolean noInvert) {
+    // It's OK to pass unlimited maxDeterminizedStates: the automaton is born small and determinized:
```
Comment (justinrsweeney) [inline solr/core/src/java/org/apache/solr/query/DynamicComplementPrefixQuery.java]: Should this comment be removed as this is not using automaton?

**Review by justinrsweeney — COMMENTED** [review]
Are there existing tests that cover this functionality with string prefix queries?

File: solr/core/src/java/org/apache/solr/query/DynamicComplementPrefixQuery.java (hunk 1 above)
Comment (magibney) [reply solr/core/src/java/org/apache/solr/query/DynamicComplementPrefixQuery.java]: thanks for catching this; fixed with e14b68661d1cdb34df51ce466e8b4673b52d32bd

**Comment by magibney** [comment]
> Are there existing tests that cover this functionality with string prefix queries?

I had assumed so (and definitely was running "informal" tests during development); but I should validate this, esp. because the existing prefix query tests are mostly on the lucene side, so it's possible that if solr itself was heavily relying on Lucene for covering prefix query, then this might not adequately cover it.

I will dig a bit deeper, thanks for calling this out.

**Review by justinrsweeney — APPROVED** [review]
LGTM

I think it'd be useful to follow up on verifying if tests exist or adding tests, but I think this could be done concurrently with getting this on staging

**Comment by magibney** [comment]
>I think it'd be useful to follow up on verifying if tests exist or adding tests, but I think this could be done concurrently with getting this on staging

After checking, this definitely needs tests -- esp. because we need to exercise the boundary between higher/lower-cardinality prefix matches and complement/direct operation, which would not be explicitly covered by existing tests and would be unlikely to happen by chance. I'll make sure to have this wrapped up before deploying to prod (I have an idea of how to proceed), but will merge now to get this in staging. Thanks!

**Pull request merged**